src/transform/pre_reduce.rs

```diff
@@ -19,39 +76,85 @@ impl run::AsDef for InertDef {
+///
+/// Here's how it works:
+/// - Each definition is visited in topological order (dependencies before
+///   dependents). In the case of cycles, one will be arbitrarily selected to be
+///   first.
+/// - The definition is reduced in a [`run::Net`]
+/// - The reduced [`run::Net`] is readback into an [`ast::Net`]
+/// - The [`ast::Net`] is encoded into a [`Vec<Instruction>`]
+/// - The [`ast::Net`] is stored in the [`State`], as it will be used later.
+/// - The [`InterpretedDef`] corresponding to the definition is mutated in-place
+///   and its instructions are replaced with the generated [`Vec<Instruction>`]
+///
+/// At the end, each mutated [`ast::Net`] is placed into the [`Book`],
+/// overriding the previous one.
+struct State<'a> {
+  book: &'a Book,
+
+  host: Host,
+  max_rwts: u64,
+
+  area: &'a Heap,
+  captured_redexes: Arc<Mutex<Vec<(run::Port, run::Port)>>>,
+
+  skip: &'a dyn Fn(&str) -> bool,
+  seen: HashMap<String, SeenState>,
+
+  rewrites: Rewrites<u64>,
+}
 
-      let mut rt = run::Net::<run::Strict>::new(&area);
-      rt.boot(host.defs.get(nam).expect("No function."));
-      rt.expand();
-      rt.reduce(max_rwts as usize);
-
-      // Move interactions with inert defs back into the net redexes array
-      for def in host.defs.values() {
-        if let Some(def) = def.downcast_ref::<InertDef>() {
-          let mut stored_redexes = def.data.0.lock().unwrap();
-          rt.redexes.extend(core::mem::take(&mut *stored_redexes));
-        }
+impl<'a> State<'a> {
+  fn visit_tree(&mut self, tree: &Tree) {
+    maybe_grow(move || match tree {
+      Tree::Era | Tree::Num { .. } | Tree::Var { .. } => (),
+      Tree::Ref { nam } => {
+        self.pre_reduce(nam);
+      }
+      Tree::Ctr { lft, rgt, .. } | Tree::Op2 { lft, rgt, .. } | Tree::Mat { sel: lft, ret: rgt } => {
+        self.visit_tree(lft);
+        self.visit_tree(rgt);
       }
-      // Place the reduced net back into the def map
-      *net = host.readback(&mut rt);
+      Tree::Op1 { rgt, .. } => self.visit_tree(rgt),
+    })
+  }
+  fn visit_net(&mut self, net: &Net) {
+    self.visit_tree(&net.root);
+    for (a, b) in &net.redexes {
+      self.visit_tree(a);
+      self.visit_tree(b);
     }
-    Ok(())
+  }
+  fn pre_reduce(&mut self, nam: &str) {
+    if self.seen.contains_key(nam) || (self.skip)(nam) || self.book.get(nam).is_none() {
+      return;
+    }
+    self.seen.insert(nam.to_owned(), SeenState::Cycled);
+    // First, pre-reduce all nets referenced by this net by walking the tree
+    self.visit_net(self.book.get(nam).unwrap());
+
+    let mut rt = run::Net::<run::Strict>::new(self.area);
+    rt.boot(self.host.defs.get(nam).expect("No function."));
+    rt.expand();
+    rt.reduce(self.max_rwts as usize);
+
+    self.rewrites += rt.rwts;
+
+    // Move interactions with inert defs back into the net redexes array
+    rt.redexes.extend(core::mem::take::<Vec<_>>(self.captured_redexes.lock().unwrap().as_mut()).into_iter());
```

Comment:
Using `drain` allows the allocation to be reused:
```suggestion
    rt.redexes.extend(self.captured_redexes.lock().unwrap().drain(..));
```